sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -833,13 +869,57 @@ struct _pi_program : _pi_object {
 
 struct _pi_kernel : _pi_object {
   _pi_kernel(ze_kernel_handle_t Kernel, pi_program Program)
-      : ZeKernel{Kernel}, Program{Program} {}
+      : ZeKernel{Kernel}, Program{Program}, MemAllocs{}, SubmissionsCount{0} {}
+
+  // Returns true if kernel has indirect access, false otherwise.
+  bool hasIndirectAccess() {
+    // Currently indirect access flag is set for all kernels and there is no API
+    // to check if kernel actually indirectly access smth.
+    return true;
+  }
 
   // Level Zero function handle.
   ze_kernel_handle_t ZeKernel;
 
   // Keep the program of the kernel.
   pi_program Program;
+
+  // If kernel has indirect access we need to make a snapshot of all existing
+  // memory allocations to defer deletion of these memory allocations to the
+  // moment when kernel execution has finished.
+  // We store iterators because iterator is not invalidated by insert/delete for
+  // std::map.
+  // Why need to take a snapshot instead of just reference-counting the
+  // allocations, because picture of active allocations can change during kernel
+  // execution (new allocations can be added) and we need to know which memory
+  // allocations were retained by this kernel to release them (and don't touch
+  // new allocations) at kernel completion.
+  // Same kernel may be submitted several times and retained allocations may be
+  // different at each submission. That's why we have a set of memory
+  // allocations here and increase ref count only once even if kernel is
+  // submitted many times. We don't want to know how many times and which
+  // allocations were retained by each submission. We release all allocations
+  // in the set only when SubmissionsCount == 0.
+  struct Hash {
+    size_t operator()(
+        const std::unordered_map<void *, MemAllocRecord>::iterator &It) const {
+      return std::hash<void *>()(It->first);
+    }
+  };
+
+  std::unordered_set<std::unordered_map<void *, MemAllocRecord>::iterator, Hash>
```

Comment:
Please add a comment explaining the data structure.

Reply:
Fixed.